sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala

```diff
@@ -50,6 +50,40 @@ trait DataSourceRegister {
   def shortName(): String
 }
 
+/**
+ * Extended version of [[DataSourceRegister]] which allows a data source to
+ * select an implementing class at runtime.  This allows plugin developers to
+ * choose the correct implementation at runtime, depending on the version of
+ * Spark the user is using.
+ *
+ *  A new instance of this class will be instantiated each time a DDL call is made.
+ *
+ */
+trait DataSourceRegisterV2 {
```

Comment:
I think this works for `TableProvider`, which is used in `DataFrameReader/Writer`. But `TableCatalog` is a different API and it uses a different way to plug into Spark, so this won't work.

cc @rdblue @brkyvz do you think this is valuable?

Reply:
This is meant strictly as an improvement to the DataSourceRegister API,
which also isn't used by the catalog.